rules/DeadCode/Rector/FunctionLike/NarrowTooWideReturnTypeRector.php

```diff
@@ -0,0 +1,265 @@
+        if (rand(0, 1)) {
+            return 'text';
+        }
+
+        return 1000;
+    }
+}
+CODE_SAMPLE
+                    ,
+                    <<<'CODE_SAMPLE'
+final class SomeClass
+{
+    public function foo(): string|int
+    {
+        if (rand(0, 1)) {
+            return 'text';
+        }
+
+        return 1000;
+    }
+}
+CODE_SAMPLE
+                ),
+            ]
+        );
+    }
+
+    public function provideMinPhpVersion(): int
+    {
+        return PhpVersionFeature::NULLABLE_TYPE;
+    }
+
+    /**
+     * @return array<class-string<Node>>
+     */
+    public function getNodeTypes(): array
+    {
+        return [ClassMethod::class, Function_::class, Closure::class, ArrowFunction::class];
+    }
+
+    /**
+     * @param ClassMethod|Function_|Closure|ArrowFunction $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        $scope = ScopeFetcher::fetch($node);
+
+        if ($this->shouldSkipNode($node, $scope)) {
+            return null;
+        }
+
+        $returnStatements = $node instanceof ArrowFunction
+            ? []
+            : $this->betterNodeFinder->findReturnsScoped($node);
+        $isAlwaysTerminating = ! $this->silentVoidResolver->hasSilentVoid($node);
+
+        if ($returnStatements === [] && ! $node instanceof ArrowFunction) {
+            return null;
+        }
+
+        $phpDocInfo = $this->phpDocInfoFactory->createFromNode($node);
+        $hasReturnDocblock = (bool) $phpDocInfo?->hasByName('@return');
+
+        if ($hasReturnDocblock) {
+            $returnType = $phpDocInfo->getReturnType();
+        } else {
+            $returnType = $node->returnType;
+            Assert::isInstanceOfAny($returnType, [UnionType::class, NullableType::class]);
+            $returnType = $this->staticTypeMapper->mapPhpParserNodePHPStanType($returnType);
+        }
+
+        $actualReturnTypes = $this->collectActualReturnTypes($node, $returnStatements, $isAlwaysTerminating);
+        $newReturnType = $this->narrowReturnType($returnType, $actualReturnTypes);
+
+        if ($newReturnType === null) {
+            return null;
+        }
+
+        $node->returnType = $this->staticTypeMapper->mapPHPStanTypeToPhpParserNode(
+            $newReturnType,
+            TypeKind::RETURN
+        );
+
+        if ($hasReturnDocblock) {
+            $this->phpDocTypeChanger->changeReturnType($node, $phpDocInfo, $newReturnType);
+        }
+
+        return $node;
+    }
+
+    private function shouldSkipNode(ClassMethod|Function_|Closure|ArrowFunction $node, Scope $scope): bool
+    {
+        $returnType = $node->returnType;
+
+        if (! $returnType instanceof UnionType && ! $returnType instanceof NullableType) {
+            return true;
+        }
+
+        if ($this->hasYield($node)) {
+            return true;
+        }
+
+        foreach ($node->params as $param) {
+            if (! $param->type instanceof Node) {
+                return true;
+            }
+        }
```

Comment:
Expr returned is not always based on param, but others :), again ,use native type instead :), see this blog post for reasoning

https://getrector.com/blog/new-in-rector-015-complete-safe-and-known-type-declarations

Reply:
Again, I can't use the native type, and this skips all methods that don't have a real php type so should be safer